### DISCUSSION TITLE
Bump packaging to 22.0

### DISCUSSION THREAD

**Discussion by mgorny** [pypa/setuptools#3722] [General]
Could you please consider updating vendored `packaging` to 22.0, and either fixing setuptools to work with it or reporting issues upstream? We're unbundling dependencies in Gentoo and noticed setuptools isn't compatible with the new version:

```pytb
  File "/tmp/portage/dev-python/setuptools-65.6.3/work/setuptools-65.6.3/./setuptools/msvc.py", line 24, in <module>
    from packaging.version import LegacyVersion
ImportError: cannot import name 'LegacyVersion' from 'packaging.version' (/usr/lib/pypy3.9/site-packages/packaging/version.py)
```

**Comment by mgorny** [comment]
Ok, maybe I'll give it a shot myself and make a PR if I succeed.

**Reply by mgorny** [reply]
Ok, this is going to be hard. @jaraco, @abravalheri FWICS the first problem is that packaging-22 removes `LegacyVersion` entirely. Do you think setuptools should ban non-PEP compliant versions at this point, or perhaps… inline `LegacyVersion`?

**Reply by mgorny** [reply]
Update: I think the only remaining blocker is that `test_warn_accidental_env_marker_misconfig` tests fail because the stray "env markers" now cause an exception in packaging. Not sure if we want to try to recover from it or just promote it from warning to an error.

**Reply by abravalheri** [reply]
I created a PR for that in https://github.com/pypa/setuptools/pull/3790.

**Comment by jaraco** [comment]
The plan is to drop support for non-compliant versions. At one point, the deprecation was masked because Setuptools had vendored an older copy of packaging without the deprecation. I believe the deprecation has been in place for some time now, so it should be viable to drop support. I definitely am averse to supporting `LegacyVersion`. My preference at the time was that the packaging ecosystem should adopt Setuptools lenient versioning semantics, but we chose to go another direction, so Setuptools should follow and align.

**Reply by abravalheri** [reply]
Hi @jaraco, I am considering an update for `packging==23.0` (https://github.com/pypa/setuptools/pull/3790) in the next version of `setuptools`.

Please let me know if think we should hold back a little longer.

**Reply by jaraco** [reply]
No concerns.

**Comment by abravalheri** [comment]
This change probably has a dependence on https://github.com/pypa/setuptools/pull/2822

**Reply by mgorny** [reply]
Oh, nice, I see @jaraco's already been working on it.
